### ISSUE TITLE
Feedback homework week 1

### ISSUE THREAD

**Issue by remarcmij** [Sob7i/hyf-javascript3#1]
Hi Sobhi, here is my feedback on your week 1 homework.

**1.** When I first loaded your `index.html` in Chrome the screen went almost totally black and I couldn't see my mouse pointer any more. So the first thing I did is to comment out the line `background-color: #04031a;` in your `style.css` and try again. Now at least I could see my mouse pointer and the `<select>` element.

**2.** I always open up the Chrome Developer Tools when I review homework and I saw a number of error messages there. I think you got the startup code a bit mixed up. Just remember that all JavaScript code that does not sit inside a function definition is immediately executed. This holds true for the call you make to `fetchJSON()` in line 92. It is best to locate **_all_** your startup code in your function `main()`. To recap what function `main()` should do:

- Set up all fixed HTML elements, including and empty `<select>` element.
- Add an event listener for the `'change'` event to the `<select>` element.
- Fetch repository information by using `fetchJSON()` and adding `<option>` elements to the `<select>` statement when the response from GitHub comes back.

It is easier to oversee what is happening when you use anonymous callbacks for `fetchJSON()` directly where you call it. This would make your function `main()` look like this:

```js
function main() {

    const root = document.getElementById('root');
    const h1 = createAndAppend('h1', root, { html: 'HYF Repositories:' });
    const header = createAndAppend('div', root);
    const select = createAndAppend('select', header, { placeholder: 'Select a repository' });
    select.addEventListener("change", theListener);

    const repContainer = createAndAppend('div', root);
    repContainer.id = 'rep-container';
    const contContainer = createAndAppend('div', root);
    contContainer.id = 'cont-container';

    fetchJSON(url, function (error, data) {
        if (error !== null) {
            console.error(error.message);
        } else {
            data.forEach(repository => {
                createAndAppend('option', select, { value: repository.name, html: repository.name });
            });
        }
    });
}
```

So what you currently have in lines 82-92 goes inside function `main()`. This function needs to be executed at startup, so that is the one that you need to specify in `window.onload`:

```js
window.onload = main;
```

**2.** Your function `theListener()` should now only be called when the user changes the selection in the `<select>` element. Placing the callback functions again directly as a parameter in the corresponding `fetchJSON()` calls again improves readability. It reduces the need to look crisscross your code to find out who is calling whom.

```js
function theListener(event) {
    const urlContributors = 'https://api.github.com/repos/HackYourFuture/' + event.target.value + '/contributors';
    const urlRepositories = 'https://api.github.com/repos/HackYourFuture/' + event.target.value;

    fetchJSON(urlContributors, function (error, contData) {
        if (error !== null) {
            console.error(error.message);
        } else {
            renderCont(contData);
        }
    });

    fetchJSON(urlRepositories, function (error, repData) {
        if (error !== null) {
            console.error(error.message);
        } else {
            renderRep(repData);
        }
    });
}
``` 

**3.** In the assignment it was indicated that you should use the `contributors_url` property that your get back in the repository information from GitHub. If you follow up, your `theListener()` function should have the second `fetchJSON()` nested in the callback of the first one, like this:

```js
function theListener(event) {
    const urlRepositories = 'https://api.github.com/repos/HackYourFuture/' + event.target.value;

    fetchJSON(urlRepositories, function (error, repData) {
        if (error !== null) {
            console.error(error.message);
        } else {
            fetchJSON(repData.contributors_url, function (error, contData) {
                if (error !== null) {
                    console.error(error.message);
                } else {
                    renderRep(repData);
                    renderCont(contData);
                }
            });
        }
    });
}
```

**4.** When you name your variables, functions and arguments you should use names that do not require guessing from the people reading your code. `renderCont` probably means render contributors, so why not name is `renderContributors`. Only a few more characters to type, and no guessing needed. Using `renderRepo` (but not `renderRep`) is probably OK because most developers will understand that `repo` means "repository". I would rename the `theListener()` function to something like `onSelectChanged()`.

**5.** Please ensure that your repo includes a `.eslintrc.json` file so that ESLint can do its work and check your code for possible issues. You can find the recommended content in the fundamental on [Code Formatting](https://github.com/HackYourFuture/fundamentals/blob/master/fundamentals/code_formatting.md#eslint). If you have such a file already in your working directory, be sure to add it to your Git repo.

**6.** There are a number of unused variables in your code that ESLint would detect. You should avoid unused variables because they are confusing. For instance, in your function `renderRepo()`:

```js
function renderRepo(repoData) {
    const repoContainer = document.getElementById('rep-container');
    repoContainer.innerHTML = '';

    const list = createAndAppend('div', repoContainer, { id: 'list-container' });
    const ul = createAndAppend('ul', list, { html: '<strong> Repository information: </strong>' });
    const li1 = createAndAppend('li', ul, { html: 'Repository:' + " " + repoData.name });
    const li2 = createAndAppend('li', ul, { html: 'Fork:' + " " + repoData.forks_count });
    const li3 = createAndAppend('li', ul, { html: 'Updated:' + " " + repoData.updated_at });
}
```

the variables `li1`, `li2` and `'li3` are unused. You can/should remove them.

```js
function renderRepo(repoData) {
    const repoContainer = document.getElementById('rep-container');
    repoContainer.innerHTML = '';

    const list = createAndAppend('div', repoContainer, { id: 'list-container' });
    const ul = createAndAppend('ul', list, { html: '<strong> Repository information: </strong>' });
    createAndAppend('li', ul, { html: 'Repository:' + " " + repoData.name });
    createAndAppend('li', ul, { html: 'Fork:' + " " + repoData.forks_count });
    createAndAppend('li', ul, { html: 'Updated:' + " " + repoData.updated_at });
}
```

**7.** Your css styling has room for improvement. For instance, the avatars are too large, and the UI is not responsive. But you can improve this in your week 2 & 3 versions.

There are some other comments I could make on details but I'll wait for week 2 / 3 to mention them.

Bottom line: the main issues were startup and code organization. With some relatively minor changes it could be made to work fully. So not bad, considering the complexity of this homework, compared to previous weeks.
